### PR TITLE
Backport amcheck to 6X

### PR DESCRIPTION
**Pull request by xiong-gang**
Backport the following commit from upstream
```
commit e2c3465a9d974be17345ed1929f02f75c098dc2b
Author: Andres Freund <andres@anarazel.de>
Date:   Thu Mar 9 15:50:40 2017 -0800

    Add amcheck extension to contrib.

    This is the beginning of a collection of SQL-callable functions to
    verify the integrity of data files.  For now it only contains code to
    verify B-Tree indexes.

    This adds two SQL-callable functions, validating B-Tree consistency to
    a varying degree.  Check the, extensive, docs for details.

    The goal is to later extend the coverage of the module to further
    access methods, possibly including the heap.  Once checks for
    additional access methods exist, we'll likely add some "dispatch"
    functions that cover multiple access methods.

    Author: Peter Geoghegan, editorialized by Andres Freund
    Reviewed-By: Andres Freund, Tomas Vondra, Thomas Munro,
       Anastasia Lubennikova, Robert Haas, Amit Langote
    Discussion: CAM3SWZQzLMhMwmBqjzK+pRKXrNUZ4w90wYMUWfkeV8mZ3Debvw@mail.gmail.com
```

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: contrib/amcheck/amcheck--1.0.sql [hunk 1]
```diff
@@ -9,15 +9,15 @@
 CREATE FUNCTION bt_index_check(index regclass)
 RETURNS VOID
 AS 'MODULE_PATHNAME', 'bt_index_check'
-LANGUAGE C STRICT PARALLEL RESTRICTED;
+LANGUAGE C STRICT;
```
Comment (ashwinstar) [inline contrib/amcheck/amcheck--1.0.sql:12]: What if we define these functions as `EXECUTE ON ALL SEGMENTS` and eliminate the need for `gp_dist_random('gp_id')`?

File: contrib/amcheck/expected/check_btree.out [hunk 2]
```diff
@@ -88,3 +112,85 @@ DROP TABLE bttest_a;
...
+ 
+(1 row)
+
+SELECT bt_index_parent_check('bttest_b_ao_idx');
+ bt_index_parent_check 
+-----------------------
+ 
+(1 row)
+
+SELECT bt_index_check('bttest_a_ao_idx') from gp_dist_random('gp_id');
```
Comment (ashwinstar) [inline contrib/amcheck/expected/check_btree.out:138]: I would be interested in seeing how the output appears if multiple segments report corruption. Please can we add set with injects corruption and validates its able to detect the same. Can add test on the lines of `ao_checksum_corruption` regress test.

Comment (ashwinstar) [reply contrib/amcheck/expected/check_btree.out:138]: Also, would be better to have `gp_segment_id` add in this test query, instead of having currently empty 3 lines returned.

**Comment by ashwinstar** [comment]
Also, would be good to know why we decided to use the very initial comment for back-port and not some other commit which mostly improved the amcheck?

**Comment by xiong-gang** [comment]
> What way are we delivering this to customers?
> Is it compiled and packaged with newer versions of GPDB6?
> Are the tests from this contrib module running in CI?

@gaos1 mentioned we are planning to compiled and packaged in the newer versions
I pushed a commit to include it in the CI, thanks.

File: contrib/amcheck/amcheck--1.0.sql (hunk 1 above)
Comment (xiong-gang) [reply contrib/amcheck/amcheck--1.0.sql:12]: It can work, the problem is the user won't be able to check the catalog index on the master in this way. An alternative way is to define two functions: one for master and one for all segments. How do you think?

File: contrib/amcheck/expected/check_btree.out (hunk 2 above)
Comment (xiong-gang) [reply contrib/amcheck/expected/check_btree.out]: I pushed a commit to copy the corrupte_file function and add a case to test it, but I only tried to corrupt the first byte, it's hard to break the link or something else in the btree structure, do you think the test is enough?

**Comment by xiong-gang** [comment]
> Also, would be good to know why we decided to use the very initial commit for back-port and not some other commit which mostly improved the amcheck?

Frankly I don't know what commit(s) should we backport, should we backport them all or should we pick some commits we think that are important?

File: contrib/amcheck/expected/check_btree.out [hunk 3]
```diff
@@ -176,21 +264,58 @@ SELECT bt_index_parent_check('bttest_b_aocs_idx');
...
+
+select corrupt_file(get_index_path('bttest_b_aocs_idx'), 0) from gp_dist_random('gp_id') where gp_segment_id = 0;
+INFO:  corrupting file base/16995/16724 at 0  (seg0 slice1 172.17.0.2:6002 pid=8283)
+CONTEXT:  PL/Python function "corrupt_file"
+ corrupt_file 
+--------------
+            0
+(1 row)
+
+SELECT gp_segment_id, bt_index_parent_check('bttest_b_aocs_idx') from gp_dist_random('gp_id');
```
Comment (ashwinstar) [inline contrib/amcheck/expected/check_btree.out:310]: So, that tells us if it fails on one segment doesn't give us idea about other segments are in good shape or not.
Means users would also have to interactively skip affected segment and as well to know if it passes on all the remaining segments or not, correct? That seems little tedious. It should be have ideally provided failed on `gp_segment_id=0` and okay for rest of segments.

I understand it might be hard to achieve that in single function model. I am more curious if problem related to index link and such corruption is encountered will it behave same way. Or will report errors properly for corrupted segments and empty results for rest of segments?

File: contrib/amcheck/amcheck--1.0.sql (hunk 1 above)
Comment (ashwinstar) [reply contrib/amcheck/amcheck--1.0.sql:12]: Users should be able to validate index anywhere required, for sure that needs to be fulfilled.
Also, though I want us to stop the usage of `gp_id` and not extend it's usage. Hence, we do need to find alternatives for it. `EXECUTE ON` syntax I feel is designed specifically for these kind of things and we should leverage that.

So, yes I feel we might have to do something similar to what we do for other things like `gp_stat_replication`. Having one function for master with `EXECUTE ON MASTER` and one for segments with `EXECUTE ON ALL SEGMENTS` and then third wrapper caller which UNION ALL on top of it. Refer to `gp_stat_replication` and friends (`gp_stat_get_master_replication()` and `gp_stat_get_segment_replication()`). I can't think of better alternative than this.

It's important though that we provide freedom to be able to verify index also on some specified random segment. I am not sure if `EXECUTE ON ALL SEGMENTS` provides us that ability or not of controlling execution to only subset of segments like you did in your corruption test to exclude `gp_segment_id=0`. If `EXECUTE ON ALL SEGMENTS` allows us to control that behaviour as well then best to leverage that.

**Comment by ashwinstar** [comment]
> Frankly I don't know what commit(s) should we backport, should we backport them all or should we pick some commits we think that are important?

I went through all the commits on PostgreSQL 10 `contrib/amcheck` and I think best would be take the latest of PostgreSQL 10 to backport instead of first commit. Most of commits are syntactical like pg_indent, .gitignore, test hardening and such. Which are all good to have so lets not skip them. I would recommend getting latest from `REL_10_STABLE` for this work

**Comment by ashwinstar** [comment]
Are we aware or tried to find out if Peter fixed any issues in btree index code in 9.4 and later versions of PostgreSQL based on cases detected by this tool? Like possibly once we roll this tool out we might get reports of issues which are already fixed (I am not much concerned for GPDB6 as its based on PostgreSQL 9.4 and we mostly should have got all the fixes for it). But for GPDB5 we need to evaluate if some fixes need to be backported or not. (Which actually is good to find outside of this work anyways)

File: contrib/amcheck/expected/check_btree.out (hunk 3 above)
Comment (xiong-gang) [reply contrib/amcheck/expected/check_btree.out]: It's a little bit hard to report on which segments the index is corrupted and on which segments the index is intact, we can try to implement a wrapper function like `pg_lock_status` to dispatch the function and catch the error from the failed segments, but I think we can do that later if there is a user request.

I haven't tried with different corruption (I am going to do that) but I think it will behave the same way because in the function definition it returns a void, so I think it will either return a void or throw an error.

File: contrib/amcheck/amcheck--1.0.sql (hunk 1 above)
Comment (xiong-gang) [reply contrib/amcheck/amcheck--1.0.sql]: I changed the interface as the way you suggested: `bt_index_check` is by default execute on any node, so user can use this function to check on specific node; `bt_index_check_on_segments` is run on all segments, can be used to check the index on user table; `bt_index_check_on_all` is a plpgsql function that will perform both functions above. As upstream `bt_index_check` returns void to indicate 'no corruption', I assume we can follow it, there's no need to return number segments empty rows. How do you think?

**Comment by xiong-gang** [comment]
> > Frankly I don't know what commit(s) should we backport, should we backport them all or should we pick some commits we think that are important?
> 
> I went through all the commits on PostgreSQL 10 `contrib/amcheck` and I think best would be take the latest of PostgreSQL 10 to backport instead of first commit. Most of commits are syntactical like pg_indent, .gitignore, test hardening and such. Which are all good to have so lets not skip them. I would recommend getting latest from `REL_10_STABLE` for this work

I cherry-picked all the commits from REL_10_STABLE except the indent one because that commit changed many other files as well.

**Comment by xiong-gang** [comment]
> Are we aware or tried to find out if Peter fixed any issues in btree index code in 9.4 and later versions of PostgreSQL based on cases detected by this tool? Like possibly once we roll this tool out we might get reports of issues which are already fixed (I am not much concerned for GPDB6 as its based on PostgreSQL 9.4 and we mostly should have got all the fixes for it). But for GPDB5 we need to evaluate if some fixes need to be backported or not. (Which actually is good to find outside of this work anyways)

Good point, I will create a story to track it. Thanks.

**Comment by ashwinstar** [comment]
> I cherry-picked all the commits from REL_10_STABLE except the indent one because that commit changed many other files as well.

You can always just take the changes for files to amcheck and ignore all other file changes. I understand it doesn't feel important functionally but tidiness of code helps in long run.

File: contrib/amcheck/expected/check_btree.out (hunk 3 above)
Comment (ashwinstar) [reply contrib/amcheck/expected/check_btree.out]: Sure, I am okay to defer and do it later if requested. Though I feel the need for it will come soon if corruptions are detected.

File: contrib/amcheck/amcheck--1.0.sql (hunk 1 above)
Comment (ashwinstar) [reply contrib/amcheck/amcheck--1.0.sql]: Looks good to me. Thank You!
